### PR TITLE
perf(table): sparse mask_to_offset_ranges for RangeWithBitmap segments

### PR DESCRIPTION
**Pull request by LeoReeYang**
## Background

With stable row ids, a row keeps its id while its physical offset can move (update / compaction), so reads must resolve `row id -> offset` through each fragment's `RowIdSequence`.

For `RangeWithBitmap` segments, `RowIdSequence::mask_to_offset_ranges` materialized the **entire id range** into a `RowAddrTreeMap`, removed holes one by one, and only then applied the mask — even when the caller asked for a single id. Cost scaled with the segment span, not with the number of requested rows. This sits on the `FilteredRead` path (`plan_scan` prefilter, `plan_batch_from_row_ids`, stable index routing), e.g. a scalar-index point lookup returning 1 row still scanned a ~1M-wide segment.

## Change

In the `RangeWithBitmap` branch, when the mask is a finite allow-list, walk the selected ids directly and accumulate bitmap prefix popcounts incrementally (`row_addrs()` yields sorted addresses). Masks without finite cardinality (e.g. full-fragment markers) keep the old range-based path. Across the tested selection densities and benchmark configuration, the per-id path showed no measurable regression, so no density threshold is introduced. Also adds boundary tests (cross-byte holes, empty/missing selections, preceding-segment offsets, full-fragment mask), a brute-force oracle plus a seeded randomized equivalence test over mixed segment types, a deterministic multi-bitmap-segment test, and two micro-benchmarks (`bitmap_mask_density`, `broad_mask_fragment`).

Note: like the rest of the implementation, offsets are grouped per segment, so ranges adjacent across a segment boundary stay split (e.g. `0..5, 5..1025`); the new tests lock this behavior in.

Update: with more than one bitmap segment, the sorted selection is now materialized once and each segment binary-searches its own id range, so a dataset-wide selection is no longer rescanned per segment (O(segments x selected) → O(selected + segments log selected)). Single-bitmap-segment sequences still stream the selection with no copy, which the density numbers above confirm.

Update 2: per-id work grows with the selection while the range path grows with the bitmap span, so selections broader than the span keep the old range-based path (`num_selected <= bitmap_span` per call). This is a heuristic, not a proof: it keeps each call near the old range-based cost instead of paying per-id scans over a dataset-wide selection in every fragment. Broad-mask oracle tests cover the fallback.

Update 3: the gatekeeper's fragment-level scenario is now pinned in the benchmarks as `broad_mask_fragment` (~10M dataset-wide allow-list on one 100k-span fragment). With the span heuristic the dataset-wide call takes ~0.91ms (old range path); with the heuristic forced off it takes ~42.5ms, confirming both the regression risk and the guard. The fragment-only mask on the same segment takes ~0.76ms via the per-id path.

## Benchmarks (x86_64 WSL, `--profile release-no-lto`)

Micro-benchmark, 1M-span bitmap segment (1/17 holes), `mask_to_offset_ranges`:

| selected ids | before | after |
| ---: | ---: | ---: |
| 1 | 3.10 ms | 5.67 µs |
| 10 | 3.16 ms | 5.43 µs |
| 1,000 | 3.28 ms | 23.6 µs |
| full (~941k) | 9.35 ms | 9.14 ms (no significant change) |

Density sweep (old TreeMap path vs per-id path, same machine/config): 1 hit 3.52ms vs 5.8µs, 10k hits 4.10ms vs 206µs, 100k hits 4.75ms vs 2.11ms, 400k hits 8.65ms vs 7.17ms, full 10.65ms vs 10.06ms (n.s.) — no crossover observed where the old path wins.

End-to-end (scratch bench, not included; 1M-row stable-row-id dataset, single fragment compacted to one bitmap segment, btree index, selective queries through `FilteredRead`):

| query | before | after |
| --- | --- | --- |
| point lookup (1 row) | 7.87 ms | 0.82 ms (-89.5%) |
| range lookup (100 rows) | 8.34 ms | 0.77 ms (-90.6%) |

## Tests

- `cargo test -p lance-table rowids`: 129 passed (includes new boundary, oracle-equivalence, and multi-segment cases)
- `cargo fmt --all -- --check`: clean
- `cargo clippy -p lance-table --all-targets` (with pre-existing `lance-io` dead_code allowed): clean

### CODE REVIEW

**Comment by LeoReeYang** [comment]
Note on the `mac-build (stable)` failure: the only failing test is `lance-index vector::kmeans::tests::test_hierarchical_kmeans_is_seeded_and_balanced` (`no empty partition`, 1 empty out of 300 clusters). This test does pure k-means training on synthetic data and never touches `mask_to_offset_ranges`; this PR only changes `lance-table/src/rowids.rs` plus a bench file, so it cannot affect that assertion. The test passes locally (macOS ARM) with this change applied. Looks like an unrelated flake / platform-sensitive statistical assertion — could a maintainer please re-run the failed job? I don't have permission to rerun upstream workflows.

**Comment by LeoReeYang** [comment]
Addressed the per-segment rescan: with more than one `RangeWithBitmap` segment, the sorted selection is now materialized once and each segment binary-searches its own id range (single-segment sequences still stream with no copy, so the reported density numbers are unchanged — re-measured 1-hit 5.6µs, full-span ~8.0ms on the same machine). Shared per-id logic was factored into `bitmap_selected_offsets`; the randomized oracle test covers multi-segment sequences on both branches. 129 rowids tests, fmt, and clippy all pass.

**Comment by LeoReeYang** [comment]
Fragment-level concern addressed: the per-id path is now taken only when `num_selected <= bitmap_span` for that call, so one `mask_to_offset_ranges` call never costs more than the old range-based path — and therefore broad queries cannot regress in total across fragments, while narrow selections keep the full sparse win (re-measured single-point 5.9µs, full-span 8.3ms, same as before). Selections without finite cardinality were already on the old path. Broad-mask cases (117 selected vs 20 span; randomized live + 2x-span out-of-range) were added to the oracle tests. 129 rowids tests, fmt, and clippy pass.

**Comment by LeoReeYang** [comment]
Fragment-level case now guarded and pinned: the per-id path is taken only when `num_selected <= bitmap_span` for that call (a heuristic holding each call near the old range-based cost, not a proof). The scenario is committed as the `broad_mask_fragment` benchmark (~10M dataset-wide allow-list on one 100k-span fragment): ~0.91ms with the guard vs ~42.5ms with it forced off, fragment-only mask ~0.76ms. Wording softened accordingly.

File: rust/lance-table/src/rowids.rs [hunk 1]
```diff
@@ -665,6 +696,51 @@ impl RowIdSequence {
                     })));
                 }
                 U64Segment::RangeWithBitmap { range, bitmap } => {
+                    // When the mask is a finite allow-list, walk the selected
+                    // ids directly instead of materializing the whole range.
+                    // `row_addrs()` yields addresses in sorted order, so bitmap
+                    // prefix counts accumulate incrementally. Density
+                    // measurements on a 1M-span segment (old TreeMap path vs.
+                    // this path: 1 hit 3.52ms vs 5.8us, 100k hits 4.75ms vs
```
Comment (BubbleCal) [inline rust/lance-table/src/rowids.rs]: remove these numbers from the comments, they may be diff on diff machines

File: rust/lance-table/benches/row_id_index.rs [hunk 2]
```diff
@@ -529,17 +530,149 @@ fn bench_shot_table(c: &mut Criterion) {
     group.finish();
 }
 
+fn bench_mask_to_offset_ranges(c: &mut Criterion) {
```
Comment (BubbleCal) [inline rust/lance-table/benches/row_id_index.rs]: remove this

File: rust/lance-table/src/rowids.rs (hunk 1 above)
Comment (LeoReeYang) [reply rust/lance-table/src/rowids.rs]: Done — absolute timings removed from the code comment; numbers live only in the PR description.

File: rust/lance-table/benches/row_id_index.rs (hunk 2 above)
Comment (LeoReeYang) [reply rust/lance-table/benches/row_id_index.rs]: Done — removed `bench_mask_to_offset_ranges`; the density sweep and `broad_mask_fragment` benches cover the same paths.

**Comment by LeoReeYang** [comment]
Note on the `linux-arm` failure: the only failing test is `lance::index::vector::ivf::v2::tests::test_legacy_ivf_pq_cosine_multivec_smoke` — IVF-PQ cosine recall 0.49 vs the 0.5 threshold (49/100 hits, one hit short). This is approximate-search recall noise on ARM, in a test whose workload was recently reduced (#8743); it exercises vector quantization, not row-id offset mapping. This PR only changes `mask_to_offset_ranges`, whose output is proven identical to a brute-force oracle across 50 seeded randomized sequences plus deterministic cases — a mapping bug would return wrong rows wholesale, not borderline recall. Unrelated to this change; retriggering CI.

**Comment by LeoReeYang** [comment]
Note on the `windows` failure: 1603 passed, 1 failed — `test_torch_index_with_nans[Legacy]` died with `torch._inductor.exc.InductorError: PermissionError: [Errno 13] Permission denied` writing torch's inductor cache file on the Windows runner. This is a CI environment issue in the Python/torch suite; this PR touches only Rust (`lance-table` row-id mapping + a Rust bench) and cannot affect torch's temp-file permissions. Retriggering.

**Review by BubbleCal — CHANGES_REQUESTED** [review]
The bitmap_segments/bitmap_span pre-scan at the top of mask_to_offset_ranges runs on every call, even when the per-id path can never be taken (BlockList/default masks where finite_allow_list is None, or sequences with no bitmap segments). That adds a fixed O(num_segments) cost to the full-scan path that wasn't there before. Could you skip the scan when there is no finite allow-list so those paths pay nothing extra?

**Review by BubbleCal — APPROVED** [review]
